Added data warehouses and transaction processing services

MoJloDou-Proger opened the pull request:
- Логика из контроллеров вынесена в классы-хранилища
- Добавлены сервисы по обработке операций
- Добавлены возможности по добавлению пользователей в друзья и выставлению фильмам "лайков"

MoJloDou-Proger committed:
Controllers and validation

MoJloDou-Proger committed:
- Валидация вынесена в класс validation.Validation, убрана вложенность if-ов
- Константы вынесены в класс constants.Constants

MoJloDou-Proger committed:
- Константы вынесены в constants.Constants
- Валидация вынесена в validation.Validation; убрана вложенность if-ов
- Тесты валидации собраны в классе ValidationTest

MoJloDou-Proger committed:
Добавлены хранилища данных и сервисы по обработке операций

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,76 @@
+@RestController
+@RequestMapping("/users")
+public class UserController{
+    UserService userService;
```
Приватный модификатор доступа не забывайте

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,28 @@
+    private String description;
+    private LocalDate releaseDate;
+    private long duration;
+    private Set<Integer> likes;
```
В целом поле likes можно было прям здесь организовать 
```
private Set<Integer> likes = new HashSet<>();
```

И тогда можно было бы добавить аннотацию ломбока - AllArgsConstructor. Но это вкусовщина, можно делать как удобнее

MoJloDou-Proger replied:
AllArgsConstructor был бы лучшим решением, однако у меня в хранилище происходит передача id, поэтому применил классический конструктор

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public List<Film> getFilms(){
+        return films;
```
Желательно оборачивать сущности базы данных (даже если они inMemory) в коллекцию обертку. Чтобы не давать доступа к сущностям самой базы.

MoJloDou-Proger replied:
Так будет действительно лучше

MoJloDou-Proger committed:
Дополнительные правки
- Модификатор UserService private
- Возвращаемые данные методов getFilms(), getUsers() обернуты в коллекцию-обёртку

MoJloDou-Proger committed:
Добавлена схема базы данных

MoJloDou-Proger committed:
Update README.md
